src/backend/env/mod.rs

```diff
@@ -1160,93 +1162,69 @@ impl State {
-            .users
-            .values()
-            .flat_map(|user| user.mintable_tokens(self, user_shares, boostraping_mode))
-        {
-            tokens_to_mint
-                .entry(user_id)
-                .and_modify(|balance| *balance += tokens)
-                .or_insert(tokens);
-        }
-
-        tokens_to_mint.retain(|user_id, _| {
-            self.users
-                .get(user_id)
-                .map(|user| user.eligible_for_minting())
-                .unwrap_or_default()
-        });
-
-        tokens_to_mint
-    }
-
-    pub fn mint(&mut self) {
-        let circulating_supply: Token = self.balances.values().sum();
-        if circulating_supply >= CONFIG.maximum_supply {
+    /// Takes the market_price (e8s per 1 token) and mints new tokens for all miners with positive
+    /// rewards according to the market price ratio.
+    pub fn mint(&mut self, market_price: u64) {
+        if market_price == 0 {
             return;
         }
 
-        let tokens_to_mint = self.tokens_to_mint();
-
-        let mut minted_tokens = 0;
-        let ratio = self.minting_ratio();
         let mut summary = Summary {
             title: "Token minting report".into(),
             description: Default::default(),
             items: Vec::default(),
         };
 
+        let miner_rewards = self
+            .users
+            .values_mut()
+            .filter(|user| user.mode == Mode::Mining)
+            .fold(HashMap::new(), |mut acc, user| {
+                let rewards = user.take_positive_rewards();
+                if rewards == 0 {
+                    return acc;
+                }
+                acc.insert(user.id, rewards);
+                acc
+            });
+
+        // burn a corresponding amount credits to generate revenue
+        self.burned_cycles += miner_rewards.values().sum::<i64>();
+
+        let tokens_to_mint: HashMap<_, _> = miner_rewards
+            .into_iter()
+            .filter_map(|(user_id, rewards)| {
+                let e8s_earned = (rewards as f64 / CONFIG.credits_per_xdr as f64
+                    * self.e8s_for_one_xdr as f64) as u64;
+                let tokens_earned = e8s_earned / market_price;
+                if tokens_earned == 0 {
+                    return None;
+                }
+                Some((user_id, tokens_earned))
+            })
+            .collect();
         let total_tokens_to_mint: u64 = tokens_to_mint.values().sum();
 
-        if ratio > 1
-            && total_tokens_to_mint * 100 / circulating_supply > CONFIG.minting_threshold_percentage
-        {
+        if total_tokens_to_mint > CONFIG.max_funding_amount {
             self.logger.warn(format!(
-                "Skipping minting: `{}` tokens exceed the configured threshold of `{}`% of existing supply.",
-                total_tokens_to_mint, CONFIG.minting_threshold_percentage
+                "Skipping minting: new tokens amount `{}` exceeds the configured threshold of `{}`",
```

Comment:
We have already taken rewards and burned cycles at this point. Should we revert those changes when skipping? Or better: check for the max funding before doing the changes?

Reply:
I actually wonder if this check is correct at all. Imagine Taggr gets slashdotted and generates a lot of revenue in a week, that would correspond to more than 1k tokens. Why would not mint them? Based on which logic?

If we won't to throttle the distribution, then maybe we should throttle the minting: i.e. as soon as we reached 1k tokens, we simply stop processing rewards and they become sort of a backlog for the next minting?

Reply:
Good point, actually! I think having an early return on reaching the threshold is indeed the safest we can do for now without impacting legitimate cases... Thanks Christian! Will change.